controllers/chaosimpl/physicalmachinechaos/impl.go

```diff
@@ -0,0 +1,202 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package physicalmachinechaos
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"go.uber.org/fx"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/controllers/common"
+)
+
+type Impl struct {
+	client.Client
+	Log logr.Logger
+}
+
+func (impl *Impl) Apply(ctx context.Context, index int, records []*v1alpha1.Record, obj v1alpha1.InnerObject) (v1alpha1.Phase, error) {
+	impl.Log.Info("apply physical machine chaos")
+
+	physicalMachinechaos := obj.(*v1alpha1.PhysicalMachineChaos)
+	addresses := records[index].Id
+	addressArray := strings.Split(addresses, ",")
+
+	// for example, physicalMachinechaos.Spec.Action is 'network-delay', action is 'network', subAction is 'delay'
+	// notice: 'process' action has no subAction, set subAction to ""
+	actions := strings.SplitN(string(physicalMachinechaos.Spec.Action), "-", 2)
+	if len(actions) == 1 {
+		actions = append(actions, "")
+	} else if len(actions) != 2 {
+		err := errors.New("action invalid")
+		return v1alpha1.NotInjected, err
+	}
+	action, subAction := actions[0], actions[1]
+	physicalMachinechaos.Spec.ExpInfo.Action = subAction
+
+	/*
+		transform ExpInfo in PhysicalMachineChaos to json data required by chaosd
+		for example:
+		    ExpInfo: &ExpInfo {
+			    UID: "123",
+				Action: "cpu",
+				StressCPU: &StressCPU {
+					Load: 1,
+					Workers: 1,
+				}
+			}
+
+			transform to json data: "{\"uid\":\"123\",\"action\":\"cpu\",\"load\":1,\"workers\":1}
+	*/
+	var expInfoMap map[string]interface{}
+	expInfoBytes, _ := json.Marshal(physicalMachinechaos.Spec.ExpInfo)
+	err := json.Unmarshal(expInfoBytes, &expInfoMap)
+	if err != nil {
+		impl.Log.Error(err, "fail to unmarshal experiment info")
+		return v1alpha1.NotInjected, err
+	}
+	configKV, ok := expInfoMap[string(physicalMachinechaos.Spec.Action)].(map[string]interface{})
+	if !ok {
+		err = errors.New("transform action config to map failed")
+		impl.Log.Error(err, "")
+		return v1alpha1.NotInjected, err
+	}
+	for k, v := range configKV {
+		expInfoMap[k] = v
+	}
+	delete(expInfoMap, string(physicalMachinechaos.Spec.Action))
+
+	expInfoBytes, err = json.Marshal(expInfoMap)
+	if err != nil {
+		impl.Log.Error(err, "fail to marshal experiment info")
+		return v1alpha1.NotInjected, err
+	}
+
+	responses := make([]*http.Response, 0, len(addressArray))
+	defer func() {
+		for _, resp := range responses {
+			resp.Body.Close()
+		}
+	}()
```

Comment:
I think it would be better to abstract the http request and response into a function, and put `defer` in it. For example:
```
func startAttackRequest(){
  req, err := http.NewRequest()
  ...
  defer resp.Body.Close()
}

for _, address := range addressArray {
  ...
  startAttackRequest()
}
```

Reply:
done